lib/streamlit/state/widgets.py

```diff
@@ -182,59 +218,102 @@ def coalesce_widget_states(
     return coalesced
 
 
-class WidgetManager:
+class WidgetManager(object):
     """Stores widget values for a single connected session."""
 
     def __init__(self):
-        self._state: Dict[str, Widget] = {}
+        self._widgets: Dict[str, Widget] = {}
+        self._prev_widgets: Dict[str, Widget] = {}
 
     def __repr__(self) -> str:
         return util.repr_(self)
 
     def get_widget_value(self, widget_id: str) -> Optional[Any]:
         """Return the value of a widget, or None if no value has been set."""
-        wstate = self._state.get(widget_id, None)
-        if wstate is None:
+        widget = self._widgets.get(widget_id, None)
+        if widget is None:
             return None
 
-        return wstate.value()
+        return widget.value
 
-    def set_state(self, widget_states: WidgetStates) -> None:
-        """Copy the state from a WidgetStates protobuf into our state dict."""
-        self._state = {}
+    def get_prev_widget_value(self, widget_id: str) -> Optional[Any]:
+        prev_widget = self._prev_widgets.get(widget_id, None)
+        if prev_widget is None:
+            return None
+
+        return prev_widget.value
+
+    def mark_widgets_as_old(self) -> None:
+        self._prev_widgets = self._widgets
+        self._widgets = {}
+
+    def set_widget_states(self, widget_states: WidgetStates) -> None:
+        """Copy the state from a WidgetStates protobuf into self._widgets."""
         for wstate in widget_states.widgets:
-            widget = Widget(wstate)
-            self._state[wstate.id] = widget
+            widget = self._widgets.get(wstate.id, None)
+
+            if widget is None:
+                self._widgets[wstate.id] = Widget(wstate.id, state=wstate)
+            else:
+                widget.state = wstate
+
+    def set_widget_attrs(
+        self,
+        widget_id: str,
+        callback: Optional[WidgetCallback],
+        deserializer: WidgetDeserializer,
+        args: Optional[WidgetArgs],
+        kwargs: Optional[WidgetKwargs],
+    ) -> None:
+        widget = self._widgets.get(widget_id, None)
+        if widget is None:
+            widget = Widget(widget_id)
+            self._widgets[widget_id] = widget
+
+        if callback is not None:
+            widget.callback = callback
+            widget.callback_args = args
+            widget.callback_kwargs = kwargs
+        widget.deserializer = deserializer
+
+    def call_callbacks(self) -> None:
+        # The callbacks to run are those installed on the *previous* script
+        # run -- the ones for this run have yet to be installed.
+        for widget in self._prev_widgets.values():
+            callback = widget.callback
+            if callback is None:
+                continue
+
+            curr_value = self.get_widget_value(widget.id)
+            prev_value = self.get_prev_widget_value(widget.id)
+            if curr_value != prev_value:
```

Comment:
Since we only get the values to check if it has changed, we could pull this into a clean function

Reply:
:+1: done